docs/adr/0001-branching-strategy.md

```diff
@@ -0,0 +1,92 @@
+# Document and support branching strategy for allowing parallel development when development dependencies exist
+
+* Status: proposed
+* Deciders: [Codeowners](../../CODEOWNERS)
+* Date: 2020-03-16
+
+## Context and Problem Statement
+
+Due to a combination of temporarily having a large influx of people giving some attention to `react-uswds`, and our intention to be careful about releasing breaking changes, we recently created develop branches corresponding to [Github milestones](https://github.com/trussworks/react-uswds/milestones) parallel to `main` to allow simultaneous streams of work that would have previously been blocked (branches `V2` and `usdws-2.9.0`). 
```

Comment:
I'm still not sure why work was blocked vs creating a release with all of the current work and then starting new work with breaking changes on main. I don't feel like you've discussed that here.

As we talked about when we chatted, the developer workflow and the release workflow are necessarily tied together. I think the top level goals we have are around release workflow and developer workflow is an important consideration as to which release workflow we should choose. There might be multiple developer workflows per release workflow, so I think you might consider listing the top 1 or two per release workflow.

Make sense?

Reply:
> I'm still not sure why work was blocked vs creating a release with all of the current work and then starting new work with breaking changes on main. I don't feel like you've discussed that here.

The general consensus that I've understood onboarding and interacting with the project is to be very careful about making breaking changes, and ideally bundle multiple into a single release in an effort to be kinder to consumers. If there are disagreements with that philosophy, I would think we should have an ADR to get consensus there. I'd be interested to see @haworku and @suzubara weigh in on that since they have a longer history with the project than I do.

We talked about releases and what constitutes breaking vs non-breaking this week in the "irl" meeting, which we will create an ADR for. Maybe it makes sense to evaluate that before we make a decision on branching as well?

As far as I've understood, there has been no need or desire to change the release process. This ADR was written under that assumption. If true, then our branching options must accommodate the current process, though it is fair to consider the merit of flexibility of a branching strategy for alternative release strategies. That's covered in this ADR.

Ultimately what I'm gathering is we have 3 related, but separate, decisions to make via ADRs: 

1. > Why don't we just release breaking changes as frequently as we want? (ADR based of the concerns noted above)

2. > What constitutes a breaking change? (ADR from IRL meeting)

3. > Should we have a documented branching strategy for navigating development dependencies (This ADR)

And of course, if our answer to the first ADR is "Just release breaking changes immediately, always" then this ADR will no longer be needed.

I think it makes sense to evaluate them in that order, since each builds off of the preceding ADRs.

Reply:
It probably actually make sense to define a breaking change first, so the order I would recommend would be (with better titles):

1. What constitutes a breaking change?
2. How should breaking changes be released?
3. What should our documented branching strategy for navigating development dependencies look like?

Reply:
> I'm still not sure why work was blocked vs creating a release with all of the current work and then starting new work with breaking changes on main. I don't feel like you've discussed that here

@ahobson the answer to your question in detail-  the scenario is we had someone submit a PR for an existing issue  that required a major bump in the USWDS version in our `package.json`.  We didn't want to just merge that to main because we hadn't supported other USWDS changes from in between versions that including breaking changes from USWDS. That work was blocked for this reason and instead the change was separate branch for that uswds version.

The other scenario is we started a separate v2 branch when we had random PRs with breaking changes to our library, react-uswds (unrelated to uswds version).  That work would come in alongside many other types of changes and PRs.  As mentioned - _maybe we want to start releasing those breaking changes often and bumping major version randomly_  Look forward to a discussing that as needed. I def thought we didn't want to, but I bring that perspective to  from previous experience where we planned our major releases for our UI library in advance, always grouping a few breaking changes together. I thought  sure we might bump with uswds version but not at random for a single change. That perspective might be overkill for this project I realize now.